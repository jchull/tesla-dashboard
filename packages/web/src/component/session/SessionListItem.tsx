import React from 'react'
import moment from 'moment'
import { isoDurationToHuman } from '../../type/util'
import { ChargeListItemDetail } from './ChargeListItemDetail'
import { DriveListItemDetail } from './DriveListItemDetail'
import { types } from '@teslapp/common'
import { SessionListItemState } from '@teslapp/web/src/store/types/state'

export const SessionListItem: React.FC<SessionListItemState> = (
  props: SessionListItemState
) => {
  const displayDate = moment(props.session.start_date)
    .calendar()

  const duration = moment.duration(
    moment(props.session.end_date)
      .diff(moment(props.session.start_date))
  )
  const displayDuration = isoDurationToHuman(duration.toISOString())

  let color = '#3f6ae1'
  if (props.session.activity !== types.ActivityType.DRIVING) {
    color = props.session.fast_charger_present
      ? '#E31937'
      : '#00dc31'
  }
  const iconStyle = { color }

  return (
    <div
      className={
        props.selected
          ? `${props.session.activity.toLowerCase()} list-item selected`
          : `list-item ${props.session.activity.toLowerCase()} `
      }
      onClick={() => props.selectionHandler(props.session._id)}
    >
      <div className="row">
        <i className="material-icons"
           style={iconStyle}>
          {props.session.activity === types.ActivityType.DRIVING ? 'directions_car' : 'battery_charging_full'}
        </i>
        <div className="start">{displayDate}</div>
        <div className="end">{displayDuration}</div>
      </div>
      {props.session.activity === types.ActivityType.DRIVING ? (
        <DriveListItemDetail {...props}/>
      ) : (
        <ChargeListItemDetail {...props}/>
      )}
    </div>
  )
}
