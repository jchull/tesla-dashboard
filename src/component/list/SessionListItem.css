.list-item {
    padding: 6px;
    border-bottom: 1px solid rgb(153,154,156);
    flex-direction: row;
}

.list-item.drive h5 {
    color: rgb(102, 102, 102);
}

.list-item.drive.selected i.material-icons {
}

.list-item.drive i.material-icons {
    color: rgb(174, 176, 178);
}


.list-item.charge i.material-icons {
    color: #00b32a;
}

.list-item.charge h5 {
    color: #ffffff;
    display: block;
    width: 20px;
    height: 20px;
    border-radius: 10px;
    background: darkgreen;
    text-align: center;
    padding-top: 3px;
    box-sizing: border-box;
}

.list-item.selected {
    background: #f4f4f4;;
}

.list-item .row {
    display: flex;
    line-height: 30px;
}

.list-item .row span {
    margin-right: 6px;
}

.start {
    margin: 0 4px;
    vertical-align: bottom;
}

.end {
    flex: auto;
    text-align: end;
    margin: 0 4px;
}


.smtxt {
    padding-left: 4px;
    padding-right: 4px;
}
