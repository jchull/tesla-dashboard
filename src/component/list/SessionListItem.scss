.list-item {
  padding: 6px;
  border-bottom: 1px solid rgb(153, 154, 156);
  flex-direction: row;

  &.drive {
    h5 {
      color: rgb(102, 102, 102);
    }

    i.material-icons {
      color: rgb(174, 176, 178);
    }
  }

  &.charge {
    h5 {
      color: #ffffff;
      display: block;
      width: 20px;
      height: 20px;
      border-radius: 10px;
      text-align: center;
      padding-top: 3px;
      box-sizing: border-box;
    }
  }

  &.selected {
    background: #eeeeee;
  }

  .row {
    display: flex;
    line-height: 30px;

    span {
      margin-right: 6px;
    }
  }
}


.start {
  margin: 0 4px;
  vertical-align: bottom;
}

.end {
  flex: auto;
  text-align: end;
  margin: 0 4px;
}


.smtxt {
  padding-left: 4px;
  padding-right: 4px;
}
