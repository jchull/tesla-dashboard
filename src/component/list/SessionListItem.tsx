import React from 'react';
import './SessionListItem.css';
import {IVehicleSession} from '../../type/VehicleSession';
import moment from 'moment';
import {isDriveSession, isoDurationToHuman} from '../../type/util';
import {ChargeListItem} from './ChargeListItem';
import {IChargeSession} from '../../type/ChargeSession';
import {DriveListItem} from './DriveListItem';
import {IDriveSession} from '../../type/DriveSession';

interface SessionListItemState {
  session: IVehicleSession;
  selected: boolean;
  selectionHandler: Function;
}

export const SessionListItem: React.FC<SessionListItemState> = (props: SessionListItemState) => {

  const displayDate = moment(props.session.start_date)
      .calendar();

  const sessionType = isDriveSession(props.session) ? 'Drive' : 'Charge';

  const duration = moment.duration(moment(props.session.end_date)
      .diff(moment(props.session.start_date)));
  const displayDuration = isoDurationToHuman(duration.toISOString());

  let color = '#3f6ae1';
  if (!isDriveSession(props.session)) {
    // @ts-ignore
    color = props.session.fast_charger_present ? '#E31937' : '#00dc31';
  }
  const iconStyle = {color};

  return (
      <div className={props.selected ? `${sessionType.toLowerCase()} list-item selected` : `list-item ${sessionType.toLowerCase()} `}
           onClick={() => props.selectionHandler(props.session)}>
        <div className="row">
          <i className="material-icons"
             style={iconStyle}>{sessionType === 'Drive' ? 'directions_car' : 'battery_charging_full'}</i>
          <div className="start">{displayDate}</div>
          <div className="end">
            {displayDuration}
          </div>
        </div>
        {sessionType === 'Drive' ?
            <DriveListItem session={props.session as IDriveSession}/>
            :
            <ChargeListItem session={props.session as IChargeSession}/>
        }

      </div>
  );
};

