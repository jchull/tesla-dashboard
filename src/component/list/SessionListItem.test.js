import React                              from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import { act }                            from 'react-dom/test-utils';
import { SessionListItem }                from './SessionListItem';

let container = null;
beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});


it('renders session list item', () => {
  const session = {
    _id: 'test1',
    first: {
      odometer: 75
    },
    last: {
      odometer: 100
    }
  };

  const handler = (session) => console.log(session._id);

  act(() => {
    render(<SessionListItem selected={false}
                            session={session}
                            selectionHandler={handler}/>, container);
  });
  // expect(container.innerHTML)
  //   .toMatchInlineSnapshot();
});
