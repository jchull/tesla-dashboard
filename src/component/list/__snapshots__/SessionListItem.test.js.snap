// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders session list item 1`] = `
"<div class=\\"list-item drive \\">
  <div class=\\"row\\"><i class=\\"material-icons\\" style=\\"color: rgb(63, 106, 225);\\">directions_car</i>
    <div class=\\"start\\">Today at 4:55 PM</div>
    <div class=\\"end\\">53</div>
  </div>
  <div>
    <div class=\\"row\\">
      <div class=\\"start\\"><span>2.2 miles</span></div>
      <div class=\\"end\\">148.3 mph</div>
    </div>
  </div>
</div>"
`;
